### ISSUE TITLE
WampPost with Guzzle\Http

### ISSUE THREAD

**Issue by mainman** [voryx/WampPost#4]
Hi
As I wanted to disconnect a client, but from not from the browser, I thought of a solution with a PHP Client which make an RPC call from Apache. I have not managed to run the example of issue 42 (https://github.com/voryx/Thruway/issues/42).
I followed the advice of @mbonneau and downloaded WampPost.
My WampPost Client is like this :

``` php
...
$router = new Router();

$wp = new WampPost\WampPost($realm, $router->getLoop(), '127.0.0.1', 8181);

$router->addTransportProvider(new Thruway\Transport\InternalClientTransportProvider($wp));

$transportProvider = new RatchetTransportProvider("127.0.0.1", 9090);
$router->addTransportProvider($transportProvider);

$router->registerModules([
    new \Thruway\Authentication\AuthenticationManager(),
    new ChatInternalClient($realm)
]);

$router->addInternalClient(new JwtAuthenticationProvider([$realm], $jwtkey));

$router->start();
```

The ChatInternalClient (extends RouterModuleClient) contains the RPC register in onSessionStart. Like this:

``` php
$this->getCallee()->register($this->session, 'com.hostname.ban', [$this, 'banClient']);
```

And the banClient function:

``` php
$args = $args[0];
$this->router
    ->getSessionBySessionId($args['session'])
    ->shutdown();
```

I make the HTTP call with Guzzle/Http like this

``` php
$client = new Client('http://127.0.0.1:8181');
$request = $client->post('/call', [
    'headers' => ['Content-Type' => 'application/json'],
    'body' => '{"procedure": "com.hostname.ban"}'
]);
$response = $request->send();
```

I have no errors, nothing in my Router terminal. the ChatInternalClient is not called.
The response of the Guzzle Client call is a 404 not found.
I try to put the port in the array of the post method ('port' => 8181). Result is the same.
And I do not know how to auth this WampPost client with JwtAuthenticationProvider...
I am kinda lost...
Any help will be very very appreciated.
Thanks

P.S. : There is an error in WampPost.php line 94 :

``` php
$options = isset($json->options) && is_object($json->opitons) ? $json->options : null;

$options = isset($json->options) && is_object($json->options) ? $json->options : null;
```

**Comment by davidwdan** [comment]
Internal clients bypass authentication, so the JWT authentication should not be an issue.

If you look in the logs, is `com.hostname.ban` registered?

Also, you can simplify your router code a bit like this (assuming you're using v0.3.x):

``` PHP
$router = new Router();

$transportProvider = new RatchetTransportProvider("127.0.0.1", 9090);
$router->addTransportProvider($transportProvider);

$router->registerModules([new \Thruway\Authentication\AuthenticationManager()]);
$router->addInternalClient(new JwtAuthenticationProvider([$realm], $jwtkey));
$router->addInternalClient(new ChatInternalClient($realm));
$router->addInternalClient(new WampPost\WampPost($realm, $router->getLoop(), '127.0.0.1', 8181));

$router->start();

```

**Comment by mainman** [comment]
You literally saved my day ! Many thanks David.
The procedure is registered, and I have a 200 Ok.
But the procedure is not called and in the server terminal, nothing when I trigger the RPC call...
If I may ask you another question ? How do I pass arguments to the RPC calls.
Perhaps just like that ?...

``` json
'{"procedure": "com.hostname.ban", "args": {"foo": "bar"}}'
```

Thanks

**Comment by davidwdan** [comment]
Is it working now?

Here's an example RPC with all of the available arguments:

``` bash
curl -H "Content-Type: application/json" \
   -d '{"procedure": "com.myapp.procedure1", "args": ["Hello, world"], "argsKw": {}, "options": {} }' \
   http://127.0.0.1:8181/call
```

**Comment by mainman** [comment]
No unfortunetly, it isn't working. The banClient function linked to the RPC register isn't called.
Will try tomorrow (I am in France and it's getting late)

Thanks for the example and for your time

**Comment by davidwdan** [comment]
What version of Thruway are you using?

**Comment by davidwdan** [comment]
I was able to get your example to work with a couple of changes to your code:

The client:

``` PHP

class ChatInternalClient extends \Thruway\Module\RouterModuleClient {

  public function onSessionStart($session, $transport) {
    $session->register('com.hostname.ban', [$this, 'banClient']);
  }

  public function banClient($args) {

    $sessionId = $args[0];
    $this->router
      ->getSessionBySessionId($sessionId)
      ->shutdown();
  }
}

```

The router:

``` PHP

$router = new Router();
$realm  = 'realm1';

$transportProvider = new RatchetTransportProvider("127.0.0.1", 9090);
$router->addTransportProvider($transportProvider);

$router->registerModules([
  new \Thruway\Authentication\AuthenticationManager(),
  new ChatInternalClient($realm)
]);

$router->addInternalClient(new JwtAuthenticationProvider([$realm], 'some key'));
$router->addInternalClient(new WampPost\WampPost($realm, $router->getLoop(), '127.0.0.1', 8181));

$router->start();

```

Curl command:

``` BASH

curl -H "Content-Type: application/json" \
   -d '{"procedure": "com.hostname.ban", "args":["8996093042089375"]}' \
   http://127.0.0.1:8181/call

```

This is the result I get:

``` BASH
{"result":"SUCCESS","args":[null],"argsKw":{},"details":{}}
```

**Comment by mainman** [comment]
Hi @davidwdan 
First, I am on thruway dev-master.
I made the changes you suggested. But the Guzzle http call return nothing. Always got the 200 Ok.
But stll nothing in my terminal or in my logs. The banclient function isn't called.
By the way I have a "regular" client which run in parallel for chat messages and sessions. And I am on Windows (dev env isn't ready atm...)

**Comment by mainman** [comment]
I just execute the curl command in a git sh term. Hopefully, there is git.
Just works fine.
Now, it is this Guzzle call. But that is another story...
Thanks for all your time and advices.
You guys roll ! And Thruway is a great WAMP lib.

**Comment by davidwdan** [comment]
Excellent!  I'm glad it's working.
